src/Workspaces/CSharp/Portable/LanguageServices/CSharpSemanticFactsService.cs

```diff
@@ -296,5 +296,51 @@ public bool IsPartial(ITypeSymbol typeSymbol, CancellationToken cancellationToke
             return SpecializedCollections.SingletonEnumerable(
                 semanticModel.GetDeclaredSymbol(memberDeclaration, cancellationToken));
         }
+
+        public SymbolInfo GetSymbolInfo(SemanticModel semanticModel, SyntaxNode node, SyntaxToken token, CancellationToken cancellationToken)
+        {
+            switch (node)
+            {
+                case FromClauseSyntax fromClauseSyntax:
+                case JoinClauseSyntax joinClauseSyntax:
+                    var queryClauseSyntax = (QueryClauseSyntax)node;
+                    var qryInfo = semanticModel.GetQueryClauseInfo(queryClauseSyntax, cancellationToken);
+                    var hasCastInfo = qryInfo.CastInfo.Symbol != null;
+                    var hasOperationInfo = qryInfo.OperationInfo.Symbol != null;
+                    if (hasCastInfo && hasOperationInfo)
+                    {
+                        return token.IsKind(SyntaxKind.InKeyword) ? qryInfo.CastInfo : qryInfo.OperationInfo;
+                    }
```

Comment:
the logic here could use good commenting as to why you're doing what your'e doing.